### ISSUE TITLE
Support for loading identity-trust-anchors-pem arg in identity from file

### ISSUE THREAD

**Issue by rnsv** [linkerd/linkerd2#5446]
## Feature Request
I need a way to load the value of the arg identity-trust-anchors-pem in identity from volume mount

### What problem are you trying to solve?
Using cert manager to generate tls certs and hence need a way to load identity-trust-anchors-pem from a file

### How should the problem be solved?
Allow an option to read https://github.com/linkerd/linkerd2/blob/main/controller/cmd/identity/main.go#L42 from a file - similar to issuer credentials

What do you want to happen? Add any considered drawbacks.

### Any alternatives you've considered?

Is there another way to solve this problem that isn't as good a solution?

### How would users interact with this feature?

If you can, explain how users will be able to use this. Maybe some sample CLI
output?

**Comment by adleong** [comment]
@rnsv can you give some additional context here?  the trust anchors should very rarely change; can you tell me more about the situation where you would need to dynamically reload them?

**Comment by rnsv** [comment]
In my current installation, I have to manage multiple Kubernetes clusters and I use flux v1 (gitops) for cluster add-on automation.  In this setup, I am able to deploy all addons by configuring flux to fetch from a centralized git repo. I use certmanager for all certificate management.  I would like to leverage gitops for linkerd installation as well - with all the yaml files managed via git repo and certmanager responsible for certificates. Using a 2 step process with linkerd install might not be an ideal solution in this setup.  
I am able to complete this automation with few workarounds (all cert related) but I thought it would be nice if linkerd natively supports reading from secrets.
I would like to read the cert pem for identity something like this or as a file mounted in a predefined path
```
          image: ghcr.io/linkerd/controller:stable-2.9.1
          args:
            - identity
            - -log-level=info
            - -controller-namespace=linkerd
            - -identity-trust-domain=cluster.local
            - -identity-issuance-lifetime=24h0m0s
            - -identity-clock-skew-allowance=20s
            - -identity-trust-anchors-pem=$(LINKERD_IDENTITY_TRUST_ANCHORS_PEM)
            - -identity-scheme=kubernetes.io/tls
          env:
            - name: LINKERD_IDENTITY_TRUST_ANCHORS_PEM
              valueFrom:
                secretKeyRef:
                  name: linkerd-trust-pem
                  key: identity-trust-anchors-pem
```

**Comment by adleong** [comment]
Ah, gotcha.  That workaround where you load the trust anchors into an env variable looks nice; if that works I'd recommend sticking with that.

One reason I hesitate to natively support reading from secrets is that it gives the (incorrect) impression that the value will be reloaded if the value of the secret changes.  The nice thing about the env variable approach is that it makes it clear that this value is established at start-up only.

**Comment by smusick-teamwork** [comment]
I'm trying to do something very similar with ArgoCD and running up against the same thing.

**Comment by cpretzer** [comment]
@smusick-teamwork is the environment variable approach an option for you? I'd be interested to see if your ArgoCD integration can be implemented that way

**Comment by rnsv** [comment]
I believe, env var approach is a viable option.  But, please note that this option
```- -identity-trust-anchors-pem=$(LINKERD_IDENTITY_TRUST_ANCHORS_PEM)```
 expects the value to be base64 encoded - which means, we cannot directly load it from tls secret.
Can we provide an option in identity to accept the non encoded value?

**Comment by smusick-teamwork** [comment]
> @smusick-teamwork is the environment variable approach an option for you? I'd be interested to see if your ArgoCD integration can be implemented that way

No success so far, unfortunately.

**Comment by smusick-teamwork** [comment]
> @smusick-teamwork is the environment variable approach an option for you? I'd be interested to see if your ArgoCD integration can be implemented that way

So I was able to finally get this to work with ArgoCD using a fork of the chart. 
There were a quite a few issues along the way, some because Argo doesn't fully use Helm, just it's templating. I had to rip out all the YAML anchors for instance. 
But the trickiest bit was probably what @rnsv mentioned. Eventually I ended up storing the base64 encoded cert in a secret, double encoding it.

**Comment by cpretzer** [comment]
Thanks for the follow-up @smusick-teamwork 

Do you think that this is something that is common across CI/CD systems? Or is it unique to ArgoCD?

**Comment by smusick-teamwork** [comment]
I've only used ArgoCD myself, so I don't know how other systems cope with the Helm issues I ran into. 

But any gitops solution is going to run into the same problems I ran into trying to read the certs, (specifically the trust anchor). Since the configuration has to be in the repository, certs and keys need to be moved elsewhere. In our case, that is secrets created via cert manager or our own automated deployment process.

There seems to be a difference in the way the proxy consumes the trust anchor and the identity pods. The latter expects them to be base64 encoded while the former do not. So the trust anchor cert can be loaded fine into the proxy, but when it is base64 re-encoded in identity, it does not match.

**Comment by smusick-teamwork** [comment]
I spoke too soon. The pods were healthy but linkerd check fails and the proxy sidecar fails to inject.

```
linkerd-identity
----------------
× certificate config is valid
    IdentityContext.TrustAnchorsPem does not match ca.crt in linkerd-identity-issuer
    see https://linkerd.io/checks/#l5d-identity-cert-config-valid for hints
```

They're the same cert but loaded two different ways

**Comment by adleong** [comment]
Well I have some good news at least: that check has been removed https://github.com/linkerd/linkerd2/pull/5424

**Comment by adleong** [comment]
I've never used ArgoCD or Flux and I have very limited experience with gitops so it's hard for me to have an informed opinion here about the best way to integrate.  But I can say that Linkerd expects the trust root to remain constant with trust root rotation being a very rare and manual process.  So the way that I would expect this to work is that first you would set up cert-manager which would generate the trust root and store it in a secret.  Then you (or your CI system?) would fetch that trust root and supply it as a value to the Linkerd chart.  If Linkerd is installed with the trust root passed in as a chart value, then all of the base64 encoding should work out correctly.

But, again, I'm not familiar enough with the details of ArgoCD to know if this makes sense.

**Comment by smusick-teamwork** [comment]
> I've never used ArgoCD or Flux and I have very limited experience with gitops so it's hard for me to have an informed opinion here about the best way to integrate. But I can say that Linkerd expects the trust root to remain constant with trust root rotation being a very rare and manual process. So the way that I would expect this to work is that first you would set up cert-manager which would generate the trust root and store it in a secret. Then you (or your CD system?) would fetch that trust root and supply it as a value to the Linkerd chart. If Linkerd is installed with the trust root passed in as a chart value, then all of the base64 encoding should work out correctly.
 
> But, again, I'm not familiar enough with the details of ArgoCD to know if this makes sense.

The issue is that any helm values in ArgoCD are going to be stored in a Git Repo (in a yaml manifest for that application). Because of this, secrets are usually handled separately from other configuration. 

But if the trust root is already stored in a secret on the cluster, then there shouldn't be any need to add it to the Helm values. If the chart was reading the root from that secret, then you wouldn't have to repeat its value in the configuration. 


I'm giving up for now. 😞 
I've fallen down at the last hurdle - the actual proxy injection. While the proxy injector deployment looks fine, when it goes to add the template to a deployment it ignores the _proxy.tpl partial that I've tweaked to use the secret. 

```
- name: LINKERD2_PROXY_IDENTITY_TRUST_ANCHORS
   valueFrom:
     secretKeyRef:
       name: linkerd-identity-issuer
       key: ca.crt
```

Instead it looks for the original  `.Values.global.identityTrustAnchorsPEM` which doesn't exist, and fails.

**Comment by adleong** [comment]
Sorry to hear that you got so close but couldn't quite get it working 😢 

The only thing I can think of to add is that even though Cert-Manager stores the trust root in a secret, the trust root is a not a private value and can be safely stored in a Git Repo (unlike, for example, private keys).

**Comment by adleong** [comment]
Going to close this for now.  @rnsv or @smusick-teamwork feel free to re-open if you'd like to pick this up again.

**Comment by MatteoJoliveau** [comment]
Hello @adleong, sorry to revive an old issue but I'm facing the same problem: generating the trust anchor and issuer certificates via cert-manager, with everything managed via ArgoCD.

The problem I'm facing is related to the base64 encoding the identity controller expects for the envar `LINKERD2_IDENTITY_TRUST_ANCHORS`, that is impossible to achieve when reading the value from a Kubernetes secret.

Generating the secret outside of the cluster and injecting it is not feasable for us, as we deploy a large number of cluster in an automated way via Argo. We also cannot write back the trust root in the Git repos after cert-manager created it, for the same reason.

If only we could inject the plain PEM into `LINKERD2_IDENTITY_TRUST_ANCHORS` instead of requiring it to be base64 encoded, that would solve our issue nicely. Is there a specific reason for this decision? I'm trying to find some workaround but without much luck at the moment :c

**Comment by AustinSmart** [comment]
@adleong ```The only thing I can think of to add is that even though Cert-Manager stores the trust root in a secret, the trust root is a not a private value and can be safely stored in a Git Repo (unlike, for example, private keys).```

The issue is not that we wish for the value to remain secret, its that the "write back" of the value to the git repo would be a 2nd step in the install of linkerd. 

Additionally for those of us deploying to many clusters, we dont always wish to use the same root for every cluster, but we do wish to use the same config sourced from git.

With these two combined nuances, we end up having to manage multiple git repos (essentially one git config per cluster with the pre-created root).

The ideal solution for most GitOps flows (especially ArgoCD) would be the ability to source the root from a k8s secret. This is not to ease the renewal or change of the root. But rather to ease the initial loading of the cert.

TL;DR
This would move the linkerd deployment manifests (helm or raw) _consumption_ of the root from being the actual _implementation_ of the root to being a _reference_ of the implementation. This leaves room for essentially any GitOps  secrets management flow to provide the implementation separate from the linkerd deployment. 

Happy to discuss and clarify further!

**Comment by olix0r** [comment]
At the moment, core linkerd control plane components (namely the identity and destination controllers) need to have the trust anchor value embedded in their manifest since these components cannot rely on proxy-injector being available. It would be possible to fix this if we could rely on the proxy-injector being available before these components were created, but this would complicate installation significantly. We're definitely open to working out a better solution to this, but it's not a trivial change.

**Comment by adleong** [comment]
This has been fixed by https://github.com/linkerd/linkerd2/pull/6455 which adds a `linkerd-identity-trust-roots` ConfigMap which contains the configured trust root bundle.
